src/Functions/keyvaluepair/src/KeyValuePairExtractor.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <string>
+#include <unordered_map>
+
+namespace DB
+{
+/*
+ * Extracts key value pairs from strings. Strings do not need to be key-value pair only,
+ * it can contain "noise". The below grammar is a simplified representation of what is expected/ supported:
+ *
+ * line = (reserved_char* key_value_pair)*  reserved_char*
+ * key_value_pair = key kv_separator value
+ * key = <quoted_string> |  asciichar asciialphanumeric*
+ * kv_separator = ':'
+ * value = <quoted_string> | asciialphanum*
+ * item_delimiter = ','
+ *
+ * Both key and values accepts underscores as well. Special characters must be escaped.
+ * Control characters (key_value_pair_separator, item_delimiter, escape_character and enclosing_character) are customizable
+ * */
+template <typename R = std::unordered_map<std::string, std::string>>
```

Comment:
Do we use R different from `std::unordered_map<std::string, std::string>` or `std::unordered_map<std::string_view, std::string_view>`. if not, can we make it without this parametrization?

Reply:
Sort of. When I initially wrote this code, I assumed all extraction & escaping would be done on the lib side and later on I would just insert it on a CH column. Then I wrote unit tests for the lib and the unit tests expect the pairs to be properly escaped.

Since using `DB::ReplaceString` seems to be more optimal for escaping, I had to find a way to preserve the library unit tests & allow `DB::ReplaceString` to be used for escaping. Therefore, the output is parameterized based on the EscapingProcessor, which can be either NoOp (used by CH itself) or SimpleEscaping (used by the library unit tests).

One thing I would like to avoid is coupling the library to ClickHouse (i.e adding ClickHouse dependencies inside the key-value-pair extraction code).

From the top of my head, the options we have are:

1. **Keep it the way it is**. _I do understand it's not very readable and might be confusing/ over engineering, but it doesn't couple library with CH and unit tests can still expect a properly escaped output_.
2. **Remove parametrization, return `std::unordered_map<std::string_view, std::string_view>` by default and create a an Adapter/Decorator interface that will implement simple escaping and return `std::unordered_map<std::string, std::string>` to be used on unit tests**. _With this approach, both CH client code & library code will be simpler. No need for fancy templates and the unit test/ escaping problem will be solved by an adapter/decorator that'll only be used by unit tests._
3. **Remove parametrization and change expectations for key value pair extraction library. It'll no longer be expected to escape characters** _With this approach, both CH client code & library code will be simpler, but the library is not fully function & isolated._

Let me know if you agree with one of these or have a better suggestion

Reply:
So, we use `std::string_view` as much as possible to avoid copies (and in ClickHouse function implementation), but for unit-test it should be escaped that's why copying to `string` is required?

Reply:
Yes, it's only copied on unit tests because it needs to be escaped. When the function is invoked in CH, the output of the library is `std::unordered_map<std::string_view, std::string_view>` and then CH escapes/ copies it using `DB::ReplaceString`

Reply:
Let's just add it into the comment